### DISCUSSION TITLE
What is the usage for async jobs?

### DISCUSSION THREAD

**Discussion by Magicloud** [ivanceras/sauron#84] [Q&A] [answered]
For example, I could think of two kinds.

One, a background pulling from server for some status updates. I think `spawn_local` should be sufficient.

The other is common data fetching. Thanks for the example for that. In the example, I noticed that the fetching "will" be run in `Cmd`. But seems like `Component::update()` does not return `Cmd`. So I cannot fetching data directly in a `Component` event, and have to delegate to `Application` (via `XMSG`)?

**Comment by ivanceras** [comment, marked as answer]
@Magicloud yes is the current limitation, the Component can not fetch data directly, this is due to the fact that only `Cmd` can dispatch a `msg`. The Component can however specify some `Effects` which can contain information for fetching the data. You will then have to write a code to convert the `Effects` into a `Cmd` to do the actual fetching, then the result will be an `update` to the `Component` of the `Application`.

**Reply by Magicloud** [reply]
Just checking the doc, and noticed `with_local_async` of `Effects`. Seems like I can have the async fetching in it and get the result in next `update` event. Do I understand it correctly?

**Reply by ivanceras** [reply]
Ah, yeah, I forgot about that. You can use it as well. I should add more examples of this, like fetching data from the `Component`s instead of from just the `Application`.

**Reply by Magicloud** [reply]
I got an issue with `with_local_async`. It panics at calling `with_local_async` before previous one is solved.

I have two data to load in one `Component`. So I send two messages asking for loading in `init`. From log, I saw first message was handled, then the second one, then panic. No "loaded" messages were handled.

**Reply by ivanceras** [reply]
@Magicloud can you show some code? I think, you have to return a subsequent `Effects/Cmd` to fetch the next async data.

**Reply by Magicloud** [reply]
Here is the general code. Logs shows this exact message order being handled `Next`, `ReloadTags`, Panic. `Loaded` (and the one for `ReloadTags`) never occurred.

```rust
impl Component<Msg, app::Msg> for Tagging {
    fn init(&mut self) -> Effects<Msg, app::Msg> {
        log(&Array::of1(&"Tagging init".into()));
        Effects::with_local(vec![Msg::Next, Msg::ReloadTags])
    }

    fn update(&mut self, msg: Msg) -> Effects<Msg, app::Msg> {
        log(&Array::of1(&format!("Tagging update {msg:?}").into()));
        match msg {
            Msg::Next => {
                self.loading = true;
                Effects::with_local_async(vec![async {
// Fetching via reqwasm.
                        Msg::Loaded(result)
                    }]
                )
            }
            Msg::Loaded(x) => {
                self.loading = false;
                match x {
                    Err(e) => {
                        self.error_message = Some(ErrorMessage {
                            title: format!("{e}"),
                            detail: format!("{e:?}"),
                        });
                    }
                    Ok(x) => {
                        self.content = Some(x);
                    }
                }
                Effects::none()
            }
// Similar logic for `ReloadTags`.
        }
    }
}

**Reply by Magicloud** [reply]
Sorry, realized I made a confusing statement. "Logs shows ..." above, after `Next`, `ReloadTags` handler was not finished. It paniced during the process.

**Reply by ivanceras** [reply]
You can post the panic error message, so I can dig deep on this further.

**Reply by Magicloud** [reply]
```
App init frontend.js:685:17
Tagging init frontend.js:685:17
App update TaggingMsg(Next) frontend.js:685:17
Tagging update Next frontend.js:685:17
App update TaggingMsg(ReloadTags) frontend.js:685:17
Tagging update ReloadTags frontend.js:685:17
Uncaught RuntimeError: unreachable executed
    __wbg_adapter_25 http://localhost:4000/pkg/frontend.js:212
    real http://localhost:4000/pkg/frontend.js:193
frontend_bg.wasm:292184:1
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:292184
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:328961
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:50884
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:65108
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:246516
    <anonymous> http://localhost:4000/pkg/frontend_bg.wasm:332080
    __wbg_adapter_25 http://localhost:4000/pkg/frontend.js:212
    real http://localhost:4000/pkg/frontend.js:193
```

**Reply by Magicloud** [reply]
I reduced `ReloadTags` handler to minimal, and it still panic. Also tried to insert some log lines. Seems like nothing in the async block was touched.

```rust
            Msg::ReloadTags => {
                self.loading = true;
                Effects::with_local_async(
                    vec![async {
                        let result = try {
                            let value = Request::get("http://localhost:3000/tags")
                                .send().await?
                                .json::<Value>().await?;
                            let value: Vec<Tag> = serde_json::from_value(value)?;
                            value
                        };
                        Msg::LoadedTags(result)
                    }]
                )
            }
```

**Comment by ivanceras** [comment]
I've just added a new example albeit simplistic for fetching data at the component level.
https://github.com/ivanceras/sauron/tree/master/examples/fetch-data-component

**Reply by Magicloud** [reply]
Thanks.

Sorry, I was busy to checkout the whole thing. Just gave it a glance, so you fetched data directly in `init`, instead of a message handler. I will try it and see if it works with two parallel fetching.
